File: neet_code_all_cpp/array_hash/string_contains_all_bin_code_size_k.cpp
#include <iostream>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <string>

class Solution {
    public:
        // AC
        bool hasAllCodes(std::string s, int k) {
            int n = s.size();
            if(n < k) {
                return false;
            }

            auto str_to_bit = [](auto& bit_string) {
                int n = bit_string.size();
                int bit = 0;
                for(int i = n - 1; i >= 0; --i) {
                    if(bit_string[i] == '1') {
                        bit += 1 << (n - i - 1);
                    }
                }

                return bit;
            };

            std::unordered_map<int, int> bit_list;
            for(int bit = 0; bit < 1 <<k; ++bit) {
                bit_list[bit] = 0;
            }

            for(int i = 0; i <= (n - k); ++i) {
                auto sub_str = s.substr(i, k);
                ++bit_list[str_to_bit(sub_str)];
            }

            for(auto& [bit, n]: bit_list) {
                if(n == 0) {
                    return false;
                }
            }

            return true;
        }
};

// 模範解答
class SolutionAns {
    public:
        // AC
        bool hasAllCodes(std::string s, int k) {
            std::set<std::string> all_strings;
            int total = 1 << k;

            for(int i = 0; i + k <= s.size(); ++i) {
                all_strings.insert(s.substr(i, k));
                if(all_strings.size() == total) {
                    return true;
                }
            }

            return false;
        }
};

int main(void) {
    using Case = std::pair<std::string, int>;
    Case case_1 = {"00110110", 2};
    // => true
    Case case_2 = {"0110", 1};
    // => true
    Case case_3 = {"0110", 2};
    // => false

    Solution s_1;
    std::cout << s_1.hasAllCodes(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.hasAllCodes(case_2.first, case_2.second) << std::endl;
    std::cout << s_1.hasAllCodes(case_3.first, case_3.second) << std::endl;

    SolutionAns s_ans;
    std::cout << s_ans.hasAllCodes(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans.hasAllCodes(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans.hasAllCodes(case_3.first, case_3.second) << std::endl;
}

File: neet_code_all/src/array_hash/string_contains_all_bin_code_size_k.rs
use std::collections::HashMap;
use std::collections::HashSet;

struct Solution {}
impl Solution {
    // AC
    pub fn has_all_codes(s: String, k: i32) -> bool {
        let k = k as usize;
        let n = s.len();

        if n < k {
            return false;
        }

        let mut bit_list = HashMap::new();
        for bit in 0..1 << k {
            bit_list.insert(bit, 0);
        }
        /*
        for (bit, n) in &bit_list {
            println!("bit, n: {:#b}, {n}", bit);
        }
        */

        for i in 0..=(n - k) {
            let sub_str = &s[i..(i + k)];
            if let Some(num_bit) = bit_list.get_mut(&Self::str_to_bit(sub_str)) {
                *num_bit += 1;
            }
        }

        for (_, n) in bit_list {
            if n == 0 {
                return false;
            }
        }

        true
    }

    pub fn str_to_bit(bit_string: &str) -> i32 {
        let mut bit = 0;
        for (i, c) in bit_string.chars().rev().enumerate() {
            if c == '1' {
                bit += 1 << i;
            }
        }

        bit
    }
}

// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    // AC
    pub fn has_all_codes(s: String, k: i32) -> bool {
        let mut set = HashSet::new();
        let n = s.len();
        let k = k as usize;
        if n < k {
            return false;
        }
        let total = 1 << k;

        for i in 0..=(n - k) {
            set.insert(s[i..(i + k)].to_string());
            if set.len() == total {
                return true;
            }
        }

        false
    }
}

// Pythonの模範解答より(ワンライナー)
struct SolutionAnsPython {}
impl SolutionAnsPython {
    pub fn has_all_codes(s: String, k: i32) -> bool {
        (0..=(s.len() as i32 - k))
            .fold(HashSet::new(), |mut set, i| {
                set.insert(s[i as usize..(i as usize + k as usize)].to_string());
                set
            })
            .len()
            == 1 << k
    }
}

fn main() {
    let case_1 = ("00110110".to_string(), 2);
    // => true
    let case_2 = ("0110".to_string(), 1);
    // => true
    let case_3 = ("0110".to_string(), 2);
    // => false

    println!(
        "case_1: {}",
        Solution::has_all_codes(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        Solution::has_all_codes(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        Solution::has_all_codes(case_3.0.clone(), case_3.1)
    );

    println!(
        "case_1: {}",
        SolutionAnsCpp::has_all_codes(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::has_all_codes(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::has_all_codes(case_3.0.clone(), case_3.1)
    );

    println!(
        "case_1: {}",
        SolutionAnsPython::has_all_codes(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAnsPython::has_all_codes(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAnsPython::has_all_codes(case_3.0.clone(), case_3.1)
    );
}
